File: dsl-go/values.go
package dsl

import (
	"math/big"
	"strings"

	canonical "github.com/paradigm-terra/canonical-go"
)

// int256 bounds: [-2^255, 2^255 - 1].
var (
	int256Max = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 255), big.NewInt(1))
	int256Min = new(big.Int).Neg(new(big.Int).Lsh(big.NewInt(1), 255))
)

func inInt256Range(v *big.Int) bool {
	return v.Cmp(int256Min) >= 0 && v.Cmp(int256Max) <= 0
}

// Value is a runtime DSL value.
type Value struct {
	Kind string // int256|bool|string|bytes32|address|list|map|null
	I    *big.Int
	B    bool
	S    string
	List []Value
	Map  []mapEntry
}

type mapEntry struct {
	k string
	v Value
}

func isHex(s string) bool {
	for i := 0; i < len(s); i++ {
		c := s[i]
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
			return false
		}
	}
	return true
}

func isBytes32(s string) bool {
	return len(s) == 66 && strings.HasPrefix(s, "0x") && isHex(s[2:])
}

func classifyString(s string) string {
	if canonical.IsCanonicalAddress(s) {
		return "address"
	}
	if isBytes32(s) {
		return "bytes32"
	}
	return "string"
}

func makeString(s string) (Value, *DslError) {
	if _, err := canonical.UTF8NFCBytes(s); err != nil {
		return Value{}, rerr("STRING_UNASSIGNED")
	}
	return Value{Kind: "string", S: s}, nil
}

func makeBytes32(s string) (Value, *DslError) {
	if !isBytes32(s) {
		return Value{}, verr("BYTES32_MALFORMED")
	}
	return Value{Kind: "bytes32", S: "0x" + strings.ToLower(s[2:])}, nil
}

func makeAddress(s string) (Value, *DslError) {
	if !canonical.IsCanonicalAddress(s) {
		return Value{}, verr("ADDRESS_NONCANONICAL")
	}
	return Value{Kind: "address", S: s}, nil
}

func constValue(c ConstVal) (Value, *DslError) {
	switch c.Typ {
	case "int256":
		return Value{Kind: "int256", I: c.I}, nil
	case "bool":
		return Value{Kind: "bool", B: c.B}, nil
	case "string":
		return makeString(c.S)
	case "bytes32":
		return makeBytes32(c.S)
	case "address":
		return makeAddress(c.S)
	default:
		return Value{Kind: "null"}, nil
	}
}

// materialize converts a bound JcsValue into a typed DSL Value.
func materialize(j canonical.Value) (Value, *DslError) {
	switch v := j.(type) {
	case nil:
		return Value{Kind: "null"}, nil
	case bool:
		return Value{Kind: "bool", B: v}, nil
	case canonical.Int:
		bi, ok := new(big.Int).SetString(string(v), 10)
		if !ok {
			return Value{}, rerr("UNREPRESENTABLE")
		}
		if !inInt256Range(bi) {
			return Value{}, verr("INT256_RANGE")
		}
		return Value{Kind: "int256", I: bi}, nil
	case string:
		switch classifyString(v) {
		case "address":
			return makeAddress(v)
		case "bytes32":
			return makeBytes32(v)
		default:
			return makeString(v)
		}
	case []canonical.Value:
		out := make([]Value, 0, len(v))
		for _, it := range v {
			mv, err := materialize(it)
			if err != nil {
				return Value{}, err
			}
			out = append(out, mv)
		}
		return Value{Kind: "list", List: out}, nil
	case *canonical.Object:
		out := make([]mapEntry, 0)
		for _, k := range v.Keys() {
			cv, _ := v.Get(k)
			mv, err := materialize(cv)
			if err != nil {
				return Value{}, err
			}
			out = append(out, mapEntry{k: k, v: mv})
		}
		return Value{Kind: "map", Map: out}, nil
	default:
		return Value{}, rerr("UNREPRESENTABLE")
	}
}

func keyForm(v Value) (string, *DslError) {
	switch v.Kind {
	case "string", "address", "bytes32":
		return v.S, nil
	default:
		return "", verr("KEY_TYPE")
	}
}

func nfcBytes(s string) ([]byte, *DslError) {
	b, err := canonical.UTF8NFCBytes(s)
	if err != nil {
		return nil, rerr("STRING_UNASSIGNED")
	}
	return b, nil
}

// valuesEqual is DSL structural equality (DSL v1.1 §3.3). Mixed non-null kinds
// are a VALIDATION_ERROR/TYPE_MISMATCH; null compares with anything.
func valuesEqual(a, b Value) (bool, *DslError) {
	if a.Kind == "null" || b.Kind == "null" {
		return a.Kind == "null" && b.Kind == "null", nil
	}
	if a.Kind != b.Kind {
		return false, verr("TYPE_MISMATCH")
	}
	switch a.Kind {
	case "int256":
		return a.I.Cmp(b.I) == 0, nil
	case "bool":
		return a.B == b.B, nil
	case "string":
		ab, e := nfcBytes(a.S)
		if e != nil {
			return false, e
		}
		bb, e := nfcBytes(b.S)
		if e != nil {
			return false, e
		}
		return string(ab) == string(bb), nil
	case "bytes32", "address":
		return a.S == b.S, nil
	case "list":
		if len(a.List) != len(b.List) {
			return false, nil
		}
		for i := range a.List {
			eq, e := valuesEqual(a.List[i], b.List[i])
			if e != nil {
				return false, e
			}
			if !eq {
				return false, nil
			}
		}
		return true, nil
	case "map":
		if len(a.Map) != len(b.Map) {
			return false, nil
		}
		for _, ae := range a.Map {
			found := false
			for _, be := range b.Map {
				if be.k == ae.k {
					eq, e := valuesEqual(ae.v, be.v)
					if e != nil {
						return false, e
					}
					if !eq {
						return false, nil
					}
					found = true
					break
				}
			}
			if !found {
				return false, nil
			}
		}
		return true, nil
	}
	return false, verr("TYPE_MISMATCH")
}

File: dsl/src/values.ts
/**
 * Runtime value model and the DSL type system (Constraint DSL v1.1 §2).
 *
 * Values are produced from two sources: `{const: ...}` literals (typed at parse
 * time, see parse.ts) and variable lookups into the bound state/params (typed
 * here by `materialize`). String, address and NFC handling delegate to the
 * canonical layer so the DSL is byte-for-byte aligned with the encoding spec
 * (and inherits the Unicode 15.1 assigned-set pin that keeps NFC parity across
 * the TS / Rust / Go implementations).
 */

import {
  assertAssigned,
  isCanonicalAddress,
  toHex,
  utf8NfcBytes,
} from "@paradigm-terra/canonical";
import type { ConstType } from "./ast.js";
import { runtimeError, validationError } from "./errors.js";

export const INT256_MIN = -(2n ** 255n);
export const INT256_MAX = 2n ** 255n - 1n;

export function inInt256Range(v: bigint): boolean {
  return v >= INT256_MIN && v <= INT256_MAX;
}

export type Value =
  | { readonly kind: "int256"; readonly int: bigint }
  | { readonly kind: "bool"; readonly bool: boolean }
  | { readonly kind: "string"; readonly str: string }
  | { readonly kind: "bytes32"; readonly hex: string }
  | { readonly kind: "address"; readonly addr: string }
  | { readonly kind: "list"; readonly items: readonly Value[] }
  | { readonly kind: "map"; readonly entries: ReadonlyMap<string, Value> }
  | { readonly kind: "null" };

export type ValueKind = Value["kind"];

const BYTES32_RE = /^0x[0-9a-fA-F]{64}$/;

/**
 * Classify a JSON string literal into its DSL scalar type (DSL v1.1 §2):
 * a canonical raw TON address, a 0x-prefixed 32-byte hash, or a plain string.
 */
export function classifyString(s: string): "address" | "bytes32" | "string" {
  if (isCanonicalAddress(s)) return "address";
  if (BYTES32_RE.test(s)) return "bytes32";
  return "string";
}

/** Build a string value, enforcing the assigned-set pin and storing NFC form. */
export function makeString(s: string): Value {
  try {
    assertAssigned(s);
  } catch {
    throw runtimeError("STRING_UNASSIGNED", `string contains code point outside the Unicode 15.1 assigned set`);
  }
  return { kind: "string", str: s.normalize("NFC") };
}

export function makeBytes32(s: string): Value {
  if (!BYTES32_RE.test(s)) {
    throw validationError("BYTES32_MALFORMED", `bytes32 must be 0x + 64 hex digits, got ${JSON.stringify(s)}`);
  }
  return { kind: "bytes32", hex: `0x${s.slice(2).toLowerCase()}` };
}

export function makeAddress(s: string): Value {
  if (!isCanonicalAddress(s)) {
    throw validationError("ADDRESS_NONCANONICAL", `address must be canonical workchain:hex256, got ${JSON.stringify(s)}`);
  }
  return { kind: "address", addr: s };
}

export function makeInt256(v: bigint): Value {
  if (!inInt256Range(v)) {
    throw validationError("INT256_RANGE", `integer ${v} is outside the int256 range`);
  }
  return { kind: "int256", int: v };
}

/** Construct the runtime value for a typed `{const}` literal. */
export function constValue(constType: ConstType, value: bigint | boolean | string | null): Value {
  switch (constType) {
    case "int256":
      return makeInt256(value as bigint);
    case "bool":
      return { kind: "bool", bool: value as boolean };
    case "string":
      return makeString(value as string);
    case "bytes32":
      return makeBytes32(value as string);
    case "address":
      return makeAddress(value as string);
    case "null":
      return { kind: "null" };
  }
}

/**
 * Materialize a value read from bound state/params (a restricted-JCS value) into
 * a typed DSL Value. JSON arrays → list, JSON objects → map, integers → int256,
 * strings classified by `classifyString`. Lists/maps come only from bindings,
 * never from literals (DSL v1.1 §2).
 */
export function materialize(j: unknown): Value {
  if (j === null) return { kind: "null" };
  if (typeof j === "boolean") return { kind: "bool", bool: j };
  if (typeof j === "bigint") return makeInt256(j);
  if (typeof j === "number") {
    if (!Number.isInteger(j)) {
      throw runtimeError("NON_INTEGER", `non-integer number ${j} is not representable in the DSL`);
    }
    return makeInt256(BigInt(j));
  }
  if (typeof j === "string") {
    switch (classifyString(j)) {
      case "address":
        return makeAddress(j);
      case "bytes32":
        return makeBytes32(j);
      case "string":
        return makeString(j);
    }
  }
  if (Array.isArray(j)) {
    return { kind: "list", items: j.map((e) => materialize(e)) };
  }
  if (typeof j === "object") {
    const entries = new Map<string, Value>();
    for (const [k, v] of Object.entries(j as Record<string, unknown>)) {
      entries.set(k, materialize(v));
    }
    return { kind: "map", entries };
  }
  throw runtimeError("UNREPRESENTABLE", `value of type ${typeof j} cannot be materialized`);
}

/**
 * Canonical string form of a value when used as a map key for `contains_key`
 * (DSL v1.1 §3.1). Only string-representable kinds are valid keys; others throw
 * a type violation handled by the caller.
 */
export function keyForm(v: Value): string {
  switch (v.kind) {
    case "string":
      return v.str;
    case "address":
      return v.addr;
    case "bytes32":
      return v.hex;
    default:
      throw validationError("KEY_TYPE", `value of kind ${v.kind} cannot be used as a map key`);
  }
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

/**
 * Structural equality (DSL v1.1 §3.3). Returns a boolean, or throws a type
 * violation when two non-null operands have incompatible kinds (mixed types are
 * permitted only with `null`).
 */
export function valuesEqual(a: Value, b: Value): boolean {
  if (a.kind === "null" || b.kind === "null") {
    return a.kind === "null" && b.kind === "null";
  }
  if (a.kind !== b.kind) {
    throw validationError("TYPE_MISMATCH", `cannot compare ${a.kind} with ${b.kind}`);
  }
  switch (a.kind) {
    case "int256":
      return a.int === (b as typeof a).int;
    case "bool":
      return a.bool === (b as typeof a).bool;
    case "string":
      return bytesEqual(utf8NfcBytes(a.str), utf8NfcBytes((b as typeof a).str));
    case "bytes32":
      return a.hex === (b as typeof a).hex;
    case "address":
      return a.addr === (b as typeof a).addr;
    case "list": {
      const bl = (b as typeof a).items;
      if (a.items.length !== bl.length) return false;
      for (let i = 0; i < a.items.length; i++) {
        if (!valuesEqual(a.items[i]!, bl[i]!)) return false;
      }
      return true;
    }
    case "map": {
      const be = (b as typeof a).entries;
      if (a.entries.size !== be.size) return false;
      for (const [k, v] of a.entries) {
        const other = be.get(k);
        if (other === undefined) return false;
        if (!valuesEqual(v, other)) return false;
      }
      return true;
    }
  }
}

/** Human-readable kind for diagnostics (also used in golden-vector dumps). */
export function describe(v: Value): string {
  switch (v.kind) {
    case "int256":
      return `int256(${v.int})`;
    case "bool":
      return `bool(${v.bool})`;
    case "string":
      return `string(0x${toHex(utf8NfcBytes(v.str))})`;
    case "bytes32":
      return `bytes32(${v.hex})`;
    case "address":
      return `address(${v.addr})`;
    case "list":
      return `list[${v.items.length}]`;
    case "map":
      return `map{${v.entries.size}}`;
    case "null":
      return "null";
  }
}
